src/analyser.cpp

```diff
@@ -0,0 +1,1509 @@
+        //
+        //                 +--------+
+        //                 |   +    |
+        //        "+a" ==> |  / \   |
+        //                 | a  nil |
+        //                 +--------+
+        //
+        //                 +-------+
+        //                 |   +   |
+        //       "a+b" ==> |  / \  |
+        //                 | a   b |
+        //                 +-------+
+        //
+        //                 +-------------+
+        //                 |   +         |
+        //                 |  / \        |
+        //                 | a   +       |
+        //                 |    / \      |
+        // "a+b+c+d+e" ==> |   b   +     |
+        //                 |      / \    |
+        //                 |     c   +   |
+        //                 |        / \  |
+        //                 |       d   e |
+        //                 +-------------+
+
+        auto childCount = mathmlChildCount(node);
+
+        processNode(mathmlChildNode(node, 0), ast, astParent, component, equation);
+        processNode(mathmlChildNode(node, 1), ast->mPimpl->mLeftChild, ast, component, equation);
+
+        if (childCount >= 3) {
+            AnalyserEquationAstPtr astRightChild;
+            AnalyserEquationAstPtr tempAst;
+
+            processNode(mathmlChildNode(node, childCount - 1), astRightChild, nullptr, component, equation);
+
+            for (auto i = childCount - 2; i > 1; --i) {
+                tempAst = AnalyserEquationAst::create();
+
+                processNode(mathmlChildNode(node, 0), tempAst, nullptr, component, equation);
+                processNode(mathmlChildNode(node, i), tempAst->mPimpl->mLeftChild, tempAst, component, equation);
+
+                astRightChild->mPimpl->mParent = tempAst;
+
+                tempAst->mPimpl->mRightChild = astRightChild;
+                astRightChild = tempAst;
+            }
+
+            if (astRightChild != nullptr) {
+                astRightChild->mPimpl->mParent = ast;
+            }
+
+            ast->mPimpl->mRightChild = astRightChild;
+        }
+
+        // Assignment, and relational and logical operators.
+
+    } else if (node->isMathmlElement("eq")) {
+        // This element is used both to describe "a = b" and "a == b". We can
+        // distinguish between the two by checking its grand-parent. If it's a
+        // "math" element then it means that it is used to describe "a = b"
+        // otherwise it is used to describe "a == b". In the former case, there
+        // is nothing more we need to do since `ast` is already of
+        // AnalyserEquationAst::Type::ASSIGNMENT type.
+
+        if (!node->parent()->parent()->isMathmlElement("math")) {
+            ast->mPimpl->populate(AnalyserEquationAst::Type::EQ, astParent);
+
+            mModel->mPimpl->mNeedEqFunction = true;
+        }
+    } else if (node->isMathmlElement("neq")) {
+        ast->mPimpl->populate(AnalyserEquationAst::Type::NEQ, astParent);
+
```

Comment:
There are a lot of random extra lines everywhere ...?

Reply:
Yes, it helps me when I read code. IOW, yes, I find libCellML's codebase difficult to read at times...

Reply:
Just that I tend to put an empty line before assignment statements. That's my "logic" here, FWIW.